src/test/java/org/junitpioneer/testkit/PioneerAssertContainer.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.testkit;
+
+import java.util.AbstractMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.assertj.core.api.AbstractAssert;
+import org.assertj.core.api.Assertions;
+import org.junit.platform.engine.TestExecutionResult;
+
+public class PioneerAssertContainer {
+
+	public static class PioneerAssert extends AbstractAssert<PioneerAssert, ExecutionResults>
+			implements SingleReportEntryAssert, FailureAssert, ExceptionAssert {
+
+		public PioneerAssert(ExecutionResults actual) {
+			super(actual, PioneerAssert.class);
+		}
+
+		public static PioneerAssert assertThat(ExecutionResults actual) {
+			return new PioneerAssert(actual);
+		}
+
+		public SingleReportEntryAssert hasSingleReportEntry() {
+			isNotNull();
+
+			List<Map<String, String>> reportEntries = actual.reportEntries();
+			Assertions.assertThat(reportEntries).hasSize(1);
+			Map<String, String> reportEntry = reportEntries.get(0);
+			Assertions.assertThat(reportEntry).hasSize(1);
+
+			return this;
+		}
+
+		public FailureAssert hasSingleFailedTest() {
+			isNotNull();
+			Assertions.assertThat(actual.numberOfFailedTests()).isEqualTo(1);
+
+			return this;
+		}
+
+		public FailureAssert hasSingleFailedContainer() {
+			isNotNull();
+			Assertions.assertThat(actual.numberOfFailedContainers()).isEqualTo(1);
+
+			return this;
+		}
+
+		@Override
+		public void withKeyAndValue(String key, String value) {
+			Assertions
+					.assertThat(actual.reportEntries().get(0).entrySet().iterator().next())
+					.isEqualTo(new AbstractMap.SimpleEntry<>(key, value));
+		}
+
+		/**
+		 * This method should only be called from methods belonging to {@code FailureAssert}.
+		 */
+		private Optional<Throwable> firstFailureThrowable() {
+			return actual
+					.allEvents()
+					.failed()
+					.stream()
+					.findFirst()
+					.flatMap(first -> first.getPayload(TestExecutionResult.class))
+					.flatMap(TestExecutionResult::getThrowable);
+
+		}
+
+		@Override
+		public ExceptionAssert andHasException(Class<? extends Throwable> exceptionType) {
```

Comment:
My reasoning behind `andHasException` is that you can only get there if you already called a method that returns with the `FailureAssert` interface.

So it always looks like this:
```
assertThat(results).hasXY(...).andHasException();
```
In a purely grammatical context the `and` makes sense(?).

Reply:
To put an example to the discussion:

`assertThat(results).hasSingleFailedTest().andHasException().withMessageContaining("ne...")`

My first thought about naming was: We always compare `results` which is plural, so I asked myself if the methods should align with plural too. But I don't think so, because the default _AssertJ_ lib claims only a single result, independent if its a single value or a list or whatever. So we should go for singular method names too.

What I really like at @Michael1993 names is that he always has something in mind when choosing the names. Was the same at the publish condition of the `@ReportEntry` extension. And I can totally understand his thoughts here to. On the other hand I can see @aepfli point to align with the default _AssertJ_ which uses `withException" and I could imagine that this helps others people to use the API. While I see the use of `and` from grammatical context, I don't think we need it as there is always the `.` of the fluent API which says exactly this.